src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,247 @@
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.util.*;
+
+public class EnsemblBioMartRetriever extends FileRetriever {
+
+    private Properties properties;
+
+    public EnsemblBioMartRetriever() { }
+
+    public EnsemblBioMartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final String BIOMART_SPECIES_NAME_PLACEHOLDER = "BIOMART_SPECIES_NAME";
+    private static final String BIOMART_QUERY_ID_PLACEHOLDER = "BIOMART_QUERY_ID";
+    private static final String microarrayTypesBaseQuery = String.join("&",
+            "type=listAttributes",
+            "mart=ENSEMBL_MART_ENSEMBL",
+            "virtualSchema=default",
+            "dataset=" + BIOMART_SPECIES_NAME_PLACEHOLDER + "_gene_ensembl",
+            "interface=default",
+            "attributePage=feature_page",
+            "attributeGroup=external",
+            "attributeCollection=microarray"
+    );
+    private static final String UNIPROT_SUFFIX = "_uniprot";
+    private static final String MICROARRAY_SUFFIX = "_microarray";
+    private static final String UNIPROT_SWISSPROT_BIOMART_SEARCH_TERM = "uniprotswissprot";
+    private static final String UNIPROT_TREMBL_BIOMART_SEARCH_TERM = "uniprotsptrembl";
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws IOException - Thrown when file can't be found when during writing and by HTTPConnection class.
+     * @throws ParseException - Thrown by JSONParser when reading Species config file.
+     * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
+     * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
+     * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
+     */
+    public void downloadData() throws IOException, ParseException, InterruptedException, BioMartQueryException, HttpException {
+
+        // Create directory where BioMart files will be stored.
+        Files.createDirectories(Paths.get(this.destination));
+        setProperties(EnsemblBioMartUtil.getProperties());
+
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in bioMart format (eg: hsapiens).
+        for (String speciesName : EnsemblBioMartUtil.getSpeciesNames()) {
+            String speciesBioMartName = EnsemblBioMartUtil.getBioMartSpeciesName(speciesName);
+            logger.info("Retrieving BiomMart files for " + speciesBioMartName);
+
+            logger.info("Retrieving microarray data");
+            // Query BioMart for existing microarray 'types' (not ids) that exist for this species.
+            int initialRetryCount = 0;
+            Set<String> microarrayTypes = queryBioMart(getMicroarrayTypesQuery(speciesBioMartName), initialRetryCount);
```

Comment:
It may be better to override the `queryBioMart` method with a version that only takes the query and delegates to the existing `queryBioMart` method with an initialRetryCount of 0:

```
/**
 * This method queries BioMart using either a URL (see variable 'microarrayTypesBaseQuery) or an XML (see biomart-query.xml in resources) query.
 * It will retry up to 5 times if errors are returned from BioMart instead of data.
 * There are multiple cases (Frog, Both Yeasts, P. falciparum and D. discoideum) where the data does not exist in BioMart.
 * If/When this exception is thrown for this species, it can be ignored. Check the logs from previous runs to confirm.
 * @param queryString - String, URL/XML string that will be used to query BioMart.
 * @return - Set<String>, All lines of successful BioMart response.
 * @throws IOException - Thrown by HttpURLConnection, BufferedReader, URL classes.
 * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
 * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
 * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
 */
private Set<String> queryBioMart(String queryString)
    throws InterruptedException, HttpException, BioMartQueryException, IOException {

    int initialRetryCount = 0;
    return queryBioMart(queryString, initialRetryCount);
}
```

Reply:
Done